File: DP_programming/187. Repeated DNA Sequences.py
'''
All DNA is composed of a series of nucleotides abbreviated as A, C, G, and T, for example: "ACGAATTCCG". When studying DNA, it is sometimes useful to identify repeated sequences within the DNA.

Write a function to find all the 10-letter-long sequences (substrings) that occur more than once in a DNA molecule.

Example:

Input: s = "AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT"

Output: ["AAAAACCCCC", "CCCCCAAAAA"]

'''
class Solution:
    def findRepeatedDnaSequences(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        n = len(s)
        output = []
        dic = {}
        if n <=10:
            return output
        for i in range(n-9):
            if s[i:i+10] in dic:
                dic[s[i:i+10]]+=1
            else:
                dic[s[i:i+10]] = 1
        for each in dic:
            if dic[each] > 1:
                output.append(each)
        return output

if __name__ == '__main__':
    a = Solution
    s = "AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT"
    print (a.findRepeatedDnaSequences(a,s ))

File: 187_Repeated_DNA_Sequences.cpp
/*
All DNA is composed of a series of nucleotides abbreviated as A, C, G, and T, for example: "ACGAATTCCG". When studying DNA, it is sometimes useful to identify repeated sequences within the DNA.

Write a function to find all the 10-letter-long sequences (substrings) that occur more than once in a DNA molecule.

Example:

Input: s = "AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT"

Output: ["AAAAACCCCC", "CCCCCAAAAA"]
*/
#include <unordered_map>
#include <iostream>
#include <vector>
#include <string>
using namespace std;
class Solution {
public:
    vector<string> findRepeatedDnaSequences(string s) {
        //using hashtable to store 10 letter string and occurence
        unordered_map<string, int> occ;
        vector<string> output;
        if (s.length() <= 10)
        {
            return output;
        }
        for (int i = 0; i<=s.length() - 10;++i)
        {
            string tmp = s.substr(i, 10);
            //cout<<i<<endl;
            if (occ.find(tmp) == occ.end())
            {
                occ[tmp] =1;
            }
            else
            {
                occ[tmp] += 1;
            }
        }
        for (auto i: occ)
        {
            //cout<<i.first<<endl;
            //cout<<i.second<<endl;
            if ((i).second > 1)
            {
                output.push_back((i).first);
            }
        }
        return output;
    }
};

int main()
{
    Solution mySol;
    string n = "AAAAAAAAAAA";
    vector<string> out = mySol.findRepeatedDnaSequences(n);
    for (auto each:out)
    {
        cout<<each<<endl;
    }
    return 0;
}